### ISSUE TITLE
`react/jsx-curly-spacing` rule crashes ESLint 4.2.0

### ISSUE THREAD

**Issue by nwoltman** [yannickcr/eslint-plugin-react#1290]
It looks like something changed (or broke) in how ESLint handles schemas between 4.1.1 and 4.2.0.

When I try to run ESLint with the rule enabled like so

```
"react/jsx-curly-spacing": [2, {"when": "never", "children": true}]
```

I get the following error

```
can't resolve reference #/definitions/basicConfig from id #
can't resolve reference #/definitions/basicConfigOrBoolean from id #
Warning: [object Object]:
        Configuration for rule "react/jsx-curly-spacing" is invalid:
refVal[2] is not a function Use --force to continue.
```

[Quick link to the `react/jsx-curly-spacing` schema](https://github.com/yannickcr/eslint-plugin-react/blob/master/lib/rules/jsx-curly-spacing.js#L34)

**Update:** I'm guessing eslint/eslint#8852 was the culprit.

**Comment by jseminck** [comment]
Running tests today also causes an issue. At first I thought it was related to me being on a Windows machine (haven't been on a Windows machine in a while), but I guess not:

```js
  38) jsx-curly-spacing valid <App foo={42} {...bar} baz={ { 4: 2 } }>
{foo} { { bar: baz } }
</App>:
     Error: rule-tester:
        Configuration for rule "jsx-curly-spacing" is invalid:
refVal[2] is not a function
      at validateRuleOptions (node_modules\eslint\lib\config\config-validator.js:112:15)
      at Object.keys.forEach.id (node_modules\eslint\lib\config\config-validator.js:152:9)
      at Array.forEach (native)
      at validateRules (node_modules\eslint\lib\config\config-validator.js:151:30)
      at Object.validate (node_modules\eslint\lib\config\config-validator.js:206:5)
      at runRuleForItem (node_modules\eslint\lib\testers\rule-tester.js:331:23)
      at testValidTemplate (node_modules\eslint\lib\testers\rule-tester.js:413:28)
      at Context.RuleTester.it (node_modules\eslint\lib\testers\rule-tester.js:546:25)
```

**Comment by ljharb** [comment]
Refs are great; if they stopped working within the v4 line it's a bug with eslint.

**Comment by ljharb** [comment]
Please file a bug with them about the breaking change, and link it here?

**Comment by nwoltman** [comment]
Looks like someone already did - eslint/eslint#8908

**Comment by jseminck** [comment]
Sorry, I forgot to link it. I asked first on Gitter (totally forgot about that) but then was also suggested to raise an issue on the eslint repo.

Additionally I tried various different approaches to get the schema to pass but no luck (e.g. moving `definitions` to the root of the object).

**Comment by jseminck** [comment]
It seems like we cannot use `$ref` inside schema when it is an array.  It should work again when a fix for `ajv` is released (no error) but the warnings are just being ignored instead.

See these discussions for more info:
https://github.com/eslint/eslint/pull/8852#discussion_r125179364
https://github.com/eslint/eslint/issues/8908#issuecomment-313935248

**Comment by ljharb** [comment]
I'm not sure I understand what's wrong with the current schema. `{ $ref: '#/definitions/basicConfig' }` is just a macro for [`{ … the basicConfig object }`](https://github.com/yannickcr/eslint-plugin-react/blob/master/lib/rules/jsx-curly-spacing.js#L36-L53), is it not?

**Comment by jseminck** [comment]
@ljharb see this comment: https://github.com/eslint/eslint/pull/8852#discussion_r125179222

The schema `schema: [obj]` is being transformed internally by `ESLint` to:
```js
{
  type: 'array',
  items: [
    obj
  ],
  minItems: 0,
  maxItems: 1
}
```

Somehow this messes up the `$ref` definitions (how exactly I do not fully understand myself).

ESLint team fixed a similar issue for the `comma-dangle-rule`: https://github.com/gajus/eslint/blob/0e0ffc15251f2bb8836145939626dc858bae50a1/lib/rules/comma-dangle.js

I made PR with similar change for `eslint-plugin-react`.

**Comment by ljharb** [comment]
so how is that not an eslint bug? it seemed to work fine before.

**Comment by jseminck** [comment]
> Here is the example showing that with is-my-json-valid it was never working correctly: https://runkit.com/esp/596268082d308c00122b7d86. Both options should fail validation ("whatever" is not an allowed value) yet they pass. It seems that when is-my-json-valid cannot resolve $ref it quietly passes validation.

It was working in the sense that the validation error was being ignored.

It should have been the same with the new validator, but there is a bug with it that @epoberezkinwill fix: https://github.com/eslint/eslint/issues/8908#issuecomment-313935304

But in the end, it seems better to not have any errors that are being (silently) ignored, right?

**Comment by ljharb** [comment]
Closed in #1292.

**Comment by virgofx** [comment]
Can a new build get released with this PR included?

**Comment by planttheidea** [comment]
Seriously, this was closed nearly a month ago. Is a patch release too much to ask?

**Comment by Benno007** [comment]
Yes, currently waiting on this if possible

**Comment by lukeapage** [comment]
@yannickcr sorry to mention you but just in case you aren't aware of people waiting for a release, if you can manage one soon, it would be appreciated, thanks

**Comment by yannickcr** [comment]
@virgofx @planttheidea @Benno007 @lukeapage Released in v7.2.0, really sorry for the delay.
